### ISSUE TITLE
"Network Error: Connection reset by peer (os error 104)" error on many websites

### ISSUE THREAD

**Issue by Shnatsel** [algesten/ureq#318]
On many websites, e.g. http://zagreb.hr, ureq fails with the following error:

> Network Error: Connection reset by peer (os error 104)

There's over 11 thousand such websites in the top million (I'm using [Tranco list generated on the 3rd of February](https://tranco-list.eu/list/3G6L)).

Firefox and curl work fine.

Archive with all occurrences: [ureq-connection-reset-by-peer.tar.gz](https://github.com/algesten/ureq/files/5979013/ureq-connection-reset-by-peer.tar.gz)

Code used for testing: https://github.com/Shnatsel/rust-http-clients-smoke-test/blob/f206362f2e81521bbefb84007cdd25242f6db590/ureq-smoke-test/src/main.rs

**Comment by jsha** [comment]
Thanks for the reports! Will take a look. By the way, if you're interested in an additional target to do smoke testing against, running a curl build linked with Hyper+rustls would be awesome: https://github.com/abetterinternet/crustls/wiki/Building-curl-with-crustls-and-Hyper

**Comment by Shnatsel** [comment]
I believe the integration of rustls and hyper into curl is rather incomplete at this stage, so the results would not be particularly useful.

I already have the results for reqwest with rustls, would you like to see that? This might be interesting as another point of comparison. I'm currently using curl as a baseline and only reporting errors that did not occur in curl.

**Comment by Shnatsel** [comment]
Also, my entire testing setup is published this time around: https://github.com/Shnatsel/rust-http-clients-smoke-test

I still need to add docs to it, but it should be easy to repeat it with other clients.

**Comment by jsha** [comment]
For `zagreb.hr` specifically, the HTTPS version of the site supports a list of ciphersuites that have no overlap with rustls. According to https://www.ssllabs.com/ssltest/analyze.html?d=zagreb.hr, it supports:

TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA
TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256
TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA
TLS_RSA_WITH_3DES_EDE_CBC_SHA
TLS_RSA_WITH_AES_128_CBC_SHA
TLS_RSA_WITH_AES_128_CBC_SHA256
TLS_RSA_WITH_AES_256_CBC_SHA
TLS_RSA_WITH_AES_256_CBC_SHA256
TLS_RSA_WITH_RC4_128_MD5
TLS_RSA_WITH_RC4_128_SHA

According to RUST_LOG=trace output from the smoke test, rustls is offering these ciphersuites (corresponding with [rustls::ALL_CIPHERSUITES](https://docs.rs/rustls/0.19.0/src/rustls/suites.rs.html#443-455)):

TLS13_AES_128_GCM_SHA256,
TLS13_AES_256_GCM_SHA384,
TLS13_CHACHA20_POLY1305_SHA256,
TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,
TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,
TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256,
TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,
TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,
TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256,
TLS_EMPTY_RENEGOTIATION_INFO_SCSV,

I haven't checked all the other hosts yet, but I suspect something similar applies.

This failure to negotiate ciphersuites is probably a common error (1.1% by your measurement!) because rustls intentionally supports a small, safe set of them. But "Connection reset by peer" is not a great error message. We should detect connection reset specifically when it happens during handshaking, and add some useful information. For instance: "Connection reset by peer during TLS handshake. Sometimes this means the host doesn't support any of the same ciphersuites as rustls. See https://xxx."

**Comment by jsha** [comment]
Here's a curl commandline which should closely approximate how rustls handshakes. It reproduces "Connection reset by peer:"

```
curl -iL http://zagreb.hr --tlsv1.2 --ciphers TLS_AES_128_GCM_SHA256,TLS_AES_256_GCM_SHA384,TLS_CHACHA20_POLY1305_SHA256,ECDHE-ECDSA-AES128-SHA,ECDHE-ECDSA-AES256-SHA,ECDHE-ECDSA-CHACHA20-POLY1305,ECDHE-RSA-AES128-GCM-SHA256,ECDHE-RSA-AES256-GCM-SHA384,ECDHE-RSA-CHACHA20-POLY1305,RENEGOTIATION-INFO
```

Note that curl seems to ignore unrecognized cipher names, so I can't be positive I got all the names right. I copied them from curl/docs/CIPHERS.md.

**Comment by troelsarvin** [comment]
This is quite serious. I'm seeing many websites which ureq cannot connect to due to this.
One example site is https://kuhs.hvidovre.dk/ which has a perfect score in https://www.ssllabs.com/ssltest/

The following Rust application returns "Connection Failed: Connection reset by peer (os error 104)":

```rust
use ureq::Error;

fn main() {
    let url = "https://kuhs.hvidovre.dk/";
    let res = match ureq::get(url).call() {
        Ok(_) => "ok".to_string(),
        Err(Error::Status(code, _)) => code.to_string(),
        Err(Error::Transport(transport)) => transport.to_string(),
    };
    println!("res for {}: {}", url, res);
}
```

When I compare Wireshark traffic between the Rust app and running "openssl s_client -connect kuhs.hvidovre.dk:443", I notice a difference in the client hello TLS step:

Rust:
```
    TLSv1 Record Layer: Handshake Protocol: Client Hello
        Content Type: Handshake (22)
        Version: TLS 1.0 (0x0301)
        Length: 238
        Handshake Protocol: Client Hello
            Handshake Type: Client Hello (1)
            Length: 234
            Version: TLS 1.2 (0x0303)
```

Openssl's s_client:
```
    TLSv1.2 Record Layer: Handshake Protocol: Client Hello
        Content Type: Handshake (22)
        Version: TLS 1.0 (0x0301)
        Length: 305
        Handshake Protocol: Client Hello
            Handshake Type: Client Hello (1)
            Length: 301
            Version: TLS 1.2 (0x0303)
```
I don't know what to make of the difference, but there clearly is a difference.

**Comment by Shnatsel** [comment]
> One example site is https://kuhs.hvidovre.dk/ which has a perfect score in https://www.ssllabs.com/ssltest/

The "perfect score" summary at the top is misleading. All listed cipher suites are flagged WEAK by https://www.ssllabs.com/ssltest/, which is why `rustls` refuses to connect.

Support for weak cipher suites is an explicit non-goal for `rustls`. The only thing `ureq` can do here is support alternative TLS implementations that sacrifice security for compatibility, see #319.

**Comment by algesten** [comment]
This is now solved in main and will be released in ureq 2.4.0.

Default rustls is not allowing weak ciphers.

```
$ cargo run --example cureq --all-features http://zagreb.hr/
http://zagreb.hr/: Connection Failed: tls connection init failed: Connection reset by peer (os error 54)
```

With the `native-tls` feature we are most likely picking up OpenSSL (or whatever the host provides).

```
$ cargo run --example cureq --all-features -- --native-tls http://zagreb.hr/
… works
```
